File: Python3/1201 - 1400/1234 Replace the Substring for Balanced String.py
# 256ms, 78.86%; 13.3MB, 100.00%
class Solution:
    def balancedString(self, s: str) -> int:
        def to_int(c: str) -> int:
            return 'QWER'.index(c)

        a, ret = [-(len(s) >> 2)] * 4, len(s)
        for c in s:
            a[to_int(c)] += 1
        a = [x if x > 0 else 0 for x in a]
        if all([x == 0 for x in a]): return 0
        i = 0
        for j, c in enumerate(s):
            index = to_int(c)
            a[index] -= 1
            if a[index] == 0 and all([x <= 0 for x in a]):
                while True:
                    tmp = to_int(s[i])
                    a[tmp], i = a[tmp] + 1, i + 1
                    if a[tmp] > 0: break
                ret = min(ret, j - i + 2)
        return ret
    

File: 1201 - 1400/1234 Replace the Substring for Balanced String.cc
// 16ms, 96.80%; 9.7MB, 100.00%
class Solution
{
public:
    int balancedString(const string &s)
    {
        int a[4] = {}, avg = s.size() >> 2, ret = s.size();
        for (auto c : s) a[toInt(c)] += 1;
        for (int i = 0; i < 4; ++i) a[i] = max(0, a[i] - avg);
        if (all_of(a, a + 4, [](int x) { return x == 0; })) return 0;
        for (int i = 0, j = 0, size = s.size(); j < size; ++j) {
            int cur = toInt(s[j]);
            a[cur] -= 1;
            if (a[cur] == 0 && all_of(a, a + 4, [](int x) { return x <= 0; })) {
                while (++a[toInt(s[i++])] <= 0) continue;
                ret = min(ret, j - i + 2);
            }
        }
        return ret;
    }

private:
    int toInt(char c)
    {
        switch (c) {
            case 'Q': return 0;
            case 'W': return 1;
            case 'E': return 2;
            case 'R': return 3;
        }
        return -1;
    }
};
